widget/textgrid.go

```diff
@@ -0,0 +1,387 @@
+package widget
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+	"strings"
+
+	"fyne.io/fyne"
+	"fyne.io/fyne/canvas"
+	"fyne.io/fyne/theme"
+)
+
+const (
+	textAreaSpaceSymbol   = '·'
+	textAreaTabSymbol     = '→'
+	textAreaNewLineSymbol = '↵'
+)
+
+var (
+	// TextGridStyleDefault is a default style for test grid cells
+	TextGridStyleDefault TextGridStyle
+	// TextGridStyleWhitespace is the style used for whitespace characters, if enabled
+	TextGridStyleWhitespace TextGridStyle
+)
+
+// define the types seperately to the var definition so the custom style API is not leaked in their instances.
+func init() {
+	TextGridStyleDefault = &CustomTextGridStyle{}
+	TextGridStyleWhitespace = &CustomTextGridStyle{FGColor: theme.ButtonColor()}
+}
+
+// TextGridCell represents a single cell in a text grid.
+// It has a rune for the text content and a style associated with it.
+type TextGridCell struct {
+	Rune  rune
+	Style TextGridStyle
+}
+
+// TextGridStyle defines a style that can be applied to a TextGrid cell.
+type TextGridStyle interface {
+	TextColor() color.Color
+	BackgroundColor() color.Color
+}
+
+// CustomTextGridStyle is a utility type for those not wanting to define their own style types.
+type CustomTextGridStyle struct {
+	FGColor, BGColor color.Color
+}
+
+// TextColor is the color a cell should use for the text.
+func (c *CustomTextGridStyle) TextColor() color.Color {
+	return c.FGColor
+}
+
+// BackgroundColor is the color a cell should use for the background.
+func (c *CustomTextGridStyle) BackgroundColor() color.Color {
+	return c.BGColor
+}
+
+// TextGrid is a monospaced grid of characters.
+// This is designed to be used by a text editor, code preview or terminal emulator.
+type TextGrid struct {
+	BaseWidget
+	Content [][]TextGridCell
+
+	LineNumbers bool
+	Whitespace  bool
```

Comment:
Additionally, you might want to consider implementing LineNumbers as an enum to support relative line numbering à la vim.

Reply:
Thanks, not sure about an enum here - most people don't even know what relative line numbers are. In vim they are implemented as two different bools, which I'd be tempted to go with so it can be addd laer if we want.